### PR TITLE
Switch to pyhive as a hive interface

### PR DESCRIPTION
**Pull request by milimetric**
To test your notebook's conda environment needs some help:
  conda install -c conda-forge cyrus-sasl
  export CPPFLAGS="${CPPFLAGS} -isystem ${CONDA_PREFIX}/include"

In production, we would apply these changes to the base conda
environment.  There's a discussion on the need for this in phabricator.

(it needs this to allow `pip install pyhive[hive]`)

Bug: T275233

### CODE REVIEW

**Comment by milimetric** [comment]
Uh... I guess #23 was automatically closed, sorry @nshahquinn, adding you back here.  I put the clarification about cyrus-sasl in the commit message.

File: wmfdata/hive.py
Comment (nshahquinn) [inline wmfdata/hive.py:24]: We should also log a message about the deprecation if this function is used, similar to https://github.com/wikimedia/wmfdata-python/blob/master/wmfdata/mariadb.py#L148

Also, can you update the `load_csv` function to use `run` instead of `run_cli`, on lines 144 and 145 below?

File: wmfdata/hive.py
Comment (nshahquinn) [inline wmfdata/hive.py]: Please use double quotes, just to stay consistent 😊

File: wmfdata/hive.py
Comment (nshahquinn) [inline wmfdata/hive.py]: It looks like the `heap_size` argument isn't used. If it's not useful to implement that feature (I'm not sure), we should mark it as deprecated and log a warning.

We should also keep the `engine` parameter, with a deprecation warning.

File: wmfdata/hive.py
Comment (nshahquinn) [inline wmfdata/hive.py]: It looks like you can't execute multiple commands this way when using a Python client. Instead, you'll want to loop through the commands, saving the output to a variable if it exists. Here's how we do it with Spark: https://github.com/wikimedia/wmfdata-python/blob/master/wmfdata/spark.py#L316

**Review by nshahquinn — CHANGES_REQUESTED** [review]

File: wmfdata/hive.py
Comment (nshahquinn) [inline wmfdata/hive.py:46]: Make sure to use this check in the `run` function.

File: wmfdata/hive.py
Comment (nshahquinn) [inline wmfdata/hive.py:41]: Could you move this deprecation warning to the `run` function?

**Review by nshahquinn — CHANGES_REQUESTED** [review]

**Comment by nshahquinn** [comment]
> To test your notebook's conda environment needs some help:
conda install -c conda-forge cyrus-sasl
export CPPFLAGS="${CPPFLAGS} -isystem ${CONDA_PREFIX}/include"
>
> In production, we would apply these changes to the base conda
environment. There's a discussion on the need for this in phabricator.

That makes sense! Can you point me to the discussion on Phabricator? I don't see it in [T275233](https://phabricator.wikimedia.org/T275233). We should probably get that done before merging this to the main branch, and definitely before releasing this in a new version.

File: wmfdata/hive.py
Comment (milimetric) [reply wmfdata/hive.py:46]: it doesn't seem to be necessary, pyhive seems to check it and print a proper error, do you think it's still useful?

File: wmfdata/hive.py
Comment (milimetric) [reply wmfdata/hive.py:41]: but "raw" isn't deprecated, it still works, no?

File: wmfdata/hive.py
Comment (milimetric) [reply wmfdata/hive.py]: hm, so that just returns the results from the last query or None.  Would we maybe just want to return an array of all the results formatted as requested?

File: wmfdata/hive.py
Comment (milimetric) [reply wmfdata/hive.py:21]: > Can you point me to the discussion on Phabricator? I don't see it in [T275233](https://phabricator.wikimedia.org/T275233). We should probably get that done before merging this to the main branch, and definitely before releasing this in a new version.

I tried to find it, and couldn't, I'm sorry.  Luca pointed me to it, I'll ask him.  But more importantly, yeah, let's get Andrew or someone to just set up the environment.

File: wmfdata/hive.py
Comment (milimetric) [reply wmfdata/hive.py]: oh, I saw single quotes in a few places and wasn't sure.  Double quotes is fine, done.

File: wmfdata/hive.py
Comment (milimetric) [reply wmfdata/hive.py]: I couldn't figure out how to set the heap_size through pyhive, I think I'm missing something fundamental.  I asked for help and nobody else knew, but I don't think we should deprecate it.  I'll just add a warning that we're not setting it.

(done on engine)

**Comment by milimetric** [comment]
> Can you point me to the discussion on Phabricator? I don't see it in T275233. We should probably get that done before merging this to the main branch, and definitely before releasing this in a new version.

I tried to find it, and couldn't, I'm sorry. Luca pointed me to it, I'll ask him. But more importantly, yeah, let's get Andrew or someone to just set up the environment.

File: wmfdata/hive.py
Comment (milimetric) [reply wmfdata/hive.py:24]: done, and done

File: wmfdata/hive.py
Comment (milimetric) [reply wmfdata/hive.py]: EDIT: yep, definitely failing, nvm my previous message:

(not true! :) ---> And I tested multiple commands, they seemed to work ok with my concatenation, did I miss something?  What commands failed for you?)

File: wmfdata/hive.py
Comment (nshahquinn) [reply wmfdata/hive.py:24]: Thanks! It would also be good if the deprecation message advises using `run` instead.

File: wmfdata/hive.py
Comment (nshahquinn) [reply wmfdata/hive.py:41]: Well, I take "deprecated" to mean that it's recommended not to be used. And the existing code has a recommendation not to use the raw format (as with the other run functions). So I want to make sure that continues to be the case.

If you think we should _keep_ the raw format, that's a valid conversation, but a separate one 😊

File: wmfdata/hive.py
Comment (nshahquinn) [reply wmfdata/hive.py]: Hmm, what did you run?

I tried running 
```
import wmfdata as wmf

hive_db = "nshahquinn"
wmf.hive.run([
    f"DROP TABLE IF EXISTS {hive_db}.wmfdata_test_1",
    f"DROP TABLE IF EXISTS {hive_db}.wmfdata_test_2",
    f"DROP TABLE IF EXISTS {hive_db}.wmfdata_test_empty",
])
```

That gave me the following error:
> Error while compiling statement: FAILED: ParseException line 1:46 missing EOF at ';' near 'wmfdata_test_1'

File: wmfdata/hive.py
Comment (nshahquinn) [reply wmfdata/hive.py:46]: Well, the error we throw is much clearer and more helpful to the user:
> OSError: You do not have Kerberos credentials. Authenticate using `kinit` or run your script as a keytab-enabled user.

Compare that to the default error:
> TTransportException: Could not start SASL: b'Error in sasl_client_start (-1) SASL(-1): generic failure: GSSAPI Error: No credentials were supplied, or the credentials were unavailable or inaccessible (No Kerberos credentials available (default cache: FILE:/tmp/krb5cc_12049))'

Besides that, we use this function with all the run functions, so keeping it makes things consistent and gives us a single function we can modify later if the Kerberos setup changes.

File: wmfdata/hive.py
Comment (nshahquinn) [reply wmfdata/hive.py]: Hmm. I did some searching and I'm starting to think that setting the heap size isn't relevant when we're not using the `hive` CLI. That setting controls the memory used by a Hadoop process, so setting it on a worker in the Hadoop cluster always makes sense. But in this case, I think it was actually controlling the memory used by the _local_ Hadoop process started by the `hive` CLI, which was a thick client. But with `beeline` and PyHive, I don't think any processing actually happens on the client, which would mean there are no client-side memory settings to configure.

So that's my suspicion, which would mean it makes sense to deprecate the option. But I definitely could be misunderstanding the architecture.

**Comment by nshahquinn** [comment]
> > Can you point me to the discussion on Phabricator? I don't see it in T275233. We should probably get that done before merging this to the main branch, and definitely before releasing this in a new version.
> 
> I tried to find it, and couldn't, I'm sorry. Luca pointed me to it, I'll ask him. But more importantly, yeah, let's get Andrew or someone to just set up the environment.

Is [T292699](https://phabricator.wikimedia.org/T292699) the task for the necessary `CPPFLAGS` change?

File: wmfdata/hive.py
Comment (nshahquinn) [reply wmfdata/hive.py]: Oh, I see, you updated your message to say it _was_ failing. When I replied yesterday, I thought you were saying the opposite :joy:

File: wmfdata/hive.py
Comment (milimetric) [reply wmfdata/hive.py]: You can still control the memory used by the cluster, I think.  So I was thinking of finding a way to do that, but you're right that heap_size wouldn't be the right name for that anyway.  You're right, deprecating.

File: wmfdata/hive.py
Comment (milimetric) [reply wmfdata/hive.py]: my bad, yeah, this is now fixed

**Comment by milimetric** [comment]
> Is T292699 the task for the necessary CPPFLAGS change?

Yes!  I'm following up on it now.

**Comment by nshahquinn** [comment]
@milimetric thanks for the follow up! I see two issues still outstanding:
1. There's one comment still open about reinstating the deprecation notice on the "raw" format.
2. I just tried installing this version in a new environment, and I got an error where Pip was unable to build the Sasl module ([pip output](https://phabricator.wikimedia.org/P19375)). This happened whether or not Cyrus-Sasl was installed. Any ideas what's going on? (And, related to that, is Cyrus-Sasl still a required dependency? Because we haven't made any provision for getting it installed yet.)

**Comment by milimetric** [comment]
> 1. There's one comment still open about reinstating the deprecation notice on the "raw" format.

I'll take care of that soon, sorry, been out sick

> 2. I just tried installing this version in a new environment, and I got an error where Pip was unable to build the Sasl module ([pip output](https://phabricator.wikimedia.org/P19375)). This happened whether or not Cyrus-Sasl was installed. Any ideas what's going on? (And, related to that, is Cyrus-Sasl still a required dependency? Because we haven't made any provision for getting it installed yet.)

@ottomata what do you think?  In theory just installing cyrus-sasl before was all I needed.

File: wmfdata/hive.py
Comment (milimetric) [reply wmfdata/hive.py:41]: I see.  I am not a user so I don't know, I'd imagine raw would be useful if you didn't want to get pandas involved, but maybe that's just always better.  I noticed somehow `pd.read_sql` from below seems faster than the raw fetch.  If it's really going to be removed in the next major release, then totally, let's deprecate (done)

Comment (nshahquinn) [reply wmfdata/hive.py:41]: I see what you mean. I'm going to resolve this comment and move the discussion to https://phabricator.wikimedia.org/T300442, to keep this conversation focused on changing the Hive backend.

**Comment by nshahquinn** [comment]
The code in the pull request is nice and ready to merge, but won't work because [T292699](https://phabricator.wikimedia.org/T292699) prevents `sasl` from installing correctly. We're blocked on that.

**Comment by ottomata** [comment]
Alright, can you try again?  I just deployed a new version of anaconda-wmf with fixes to the scrips, as well as making cyrus-sasl and pyhive[hive] explicit dependencies in anaconda-wmf.

**Comment by nshahquinn** [comment]
Okay, I tested this again on stat1006. I was able to install it and query successfully both in a new environment and in one I had created before @ottomata rolled out the fix. I also tried querying 3-4 different Hive tables, and it worked as expected.

However, I tried running the test suite (currently merged to master but not this branch) and ran into some problems with `load_csv`. I'll comment in line about those.

File: wmfdata/hive.py
Comment (nshahquinn) [inline wmfdata/hive.py]: There are a couple issues here.

1. `msktemp` returns a path like `/tmp/tmpdw4b3ubr`. That would give us `/tmp//tmp/tmpdw4b3ubr` as our `hdfs_path`.
2. When I tried running this, I got an error because `hdfs:///tmp/tmp` is owned by milimetric, probably because it was created with default permissions when testing this code. If we put the file directly in `/tmp` (we can even use the `mkstemp` path without modification for HDFS), that should avoid the problem.

Comment (milimetric) [reply wmfdata/hive.py]: good point, I just removed the '/tmp' prefix I think that fixes it.

Comment (milimetric) [reply wmfdata/hive.py]: @nshahquinn I tested this on my notebooks and it seems like it's working ok.  Let me know if there's anything else.

**Comment by ottomata** [comment]
Great, so this should be ready to merge, ya?

**Comment by milimetric** [comment]
Yes, shall we officially do the "Andrew and Dan are maintainers and can
merge" thing now?

**Comment by nshahquinn** [comment]
I just re-ran the test suite and it passed this time, so, for what it's worth, I agree—ready to merge!

> Yes, shall we officially do the "Andrew and Dan are maintainers and can merge" thing now?

Yes, please do! I'd suggest "squash and merge" (🎃&↗↑) in this case.

**Review by nshahquinn — APPROVED** [review]

**Pull request merged**
